### ISSUE TITLE
Server Allocation failed - JavaScript heap out of memory

### ISSUE THREAD

**Issue by y1ann1s85** [inertiajs/inertia#1697]
### Versions:

- `@inertiajs/core` version: 1.0.11
- `@inertiajs/vue3` version: 1.0.11
- `vue` version: 3.3.4

### Describe the problem:

  Hello. Since we migrated from Vue 2 to Vue 3 and upgraded inertia to the latest version we keep receiving multiple errors from the node server.
  I' ve seen a related issue to it: https://github.com/inertiajs/inertia/issues/1602. We completely removed usePage from our code base to identify if this was actually the problem but unfortunately the errors keep logging.
  Any ideas or suggestions on it?

`<--- Last few GCs --->

[23043:0x59e3120] 10368422 ms: Scavenge 1987.8 (2080.3) -> 1983.1 (2080.3) MB, 19.3 / 0.0 ms  (average mu = 0.830, current mu = 0.896) allocation failure 
[23043:0x59e3120] 10368477 ms: Scavenge 1989.0 (2080.3) -> 1985.5 (2080.3) MB, 19.8 / 0.0 ms  (average mu = 0.830, current mu = 0.896) allocation failure 
[23043:0x59e3120] 10368527 ms: Scavenge 1991.4 (2080.3) -> 1987.5 (2096.3) MB, 22.9 / 0.0 ms  (average mu = 0.830, current mu = 0.896) allocation failure 

<--- JS stacktrace --->

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
1: 0xb08e80 node::Abort() [node]
 2: 0xa1b70e  [node]
 3: 0xce1890 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0xce1c37 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0xe992a5  [node]
 6: 0xea8f6d v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
 7: 0xeabc6e v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [node]
 8: 0xe6cee2 v8::internal::Factory::AllocateRaw(int, v8::internal::AllocationType, v8::internal::AllocationAlignment) [node]
 9: 0xe677fc v8::internal::FactoryBase<v8::internal::Factory>::AllocateRawArray(int, v8::internal::AllocationType) [node]
10: 0xe678d5 v8::internal::FactoryBase<v8::internal::Factory>::NewFixedArrayWithFiller(v8::internal::Handle<v8::internal::Map>, int, v8::internal::Handle<v8::internal::Oddball>, v8::internal::AllocationType) [node]
11: 0x10cad13 v8::internal::Handle<v8::internal::NameDictionary> v8::internal::HashTable<v8::internal::NameDictionary, v8::internal::NameDictionaryShape>::NewInternal<v8::internal::Isolate>(v8::internal::Isolate*, int, v8::internal::AllocationType) [node]
12: 0x10cb003 v8::internal::Handle<v8::internal::NameDictionary> v8::internal::HashTable<v8::internal::NameDictionary, v8::internal::NameDictionaryShape>::EnsureCapacity<v8::internal::Isolate>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::NameDictionary>, int, v8::internal::AllocationType) [node]
13: 0x10cb066 v8::internal::Handle<v8::internal::NameDictionary> v8::internal::Dictionary<v8::internal::NameDictionary, v8::internal::NameDictionaryShape>::Add<v8::internal::Isolate>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::NameDictionary>, v8::internal::Handle<v8::internal::Name>, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyDetails, v8::internal::InternalIndex*) [node]
14: 0x10cb6ca v8::internal::BaseNameDictionary<v8::internal::NameDictionary, v8::internal::NameDictionaryShape>::Add(v8::internal::Isolate*, v8::internal::Handle<v8::internal::NameDictionary>, v8::internal::Handle<v8::internal::Name>, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyDetails, v8::internal::InternalIndex*) [node]
15: 0x106caf3 v8::internal::JSObject::SetNormalizedProperty(v8::internal::Handle<v8::internal::JSObject>, v8::internal::Handle<v8::internal::Name>, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyDetails) [node]
16: 0x109844d v8::internal::LookupIterator::TransitionToAccessorPair(v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes) [node]
17: 0x1070a5d v8::internal::JSObject::DefineAccessor(v8::internal::LookupIterator*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes) [node]
18: 0x10712b4 v8::internal::JSReceiver::ValidateAndApplyPropertyDescriptor(v8::internal::Isolate*, v8::internal::LookupIterator*, bool, v8::internal::PropertyDescriptor*, v8::internal::PropertyDescriptor*, v8::Maybe<v8::internal::ShouldThrow>, v8::internal::Handle<v8::internal::Name>) [node]
19: 0x107178e v8::internal::JSReceiver::OrdinaryDefineOwnProperty(v8::internal::LookupIterator*, v8::internal::PropertyDescriptor*, v8::Maybe<v8::internal::ShouldThrow>) [node]
20: 0x1071ad8 v8::internal::JSReceiver::OrdinaryDefineOwnProperty(v8::internal::Isolate*, v8::internal::Handle<v8::internal::JSObject>, v8::internal::PropertyKey const&, v8::internal::PropertyDescriptor*, v8::Maybe<v8::internal::ShouldThrow>) [node]
21: 0x1071b5b v8::internal::JSReceiver::DefineOwnProperty(v8::internal::Isolate*, v8::internal::Handle<v8::internal::JSReceiver>, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyDescriptor*, v8::Maybe<v8::internal::ShouldThrow>) [node]
22: 0x1071cb8 v8::internal::JSReceiver::DefineProperty(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>) [node]
23: 0xd652b2 v8::internal::Builtin_ObjectDefineProperty(int, unsigned long*, v8::internal::Isolate*) [node]
24: 0x15d9cf9  [node]`

**Comment by craigrileyuk** [comment]
SSR/Node compilation is a lot less forgiving with your code than the browser and it's possible that previous version of your software were recovering from a fault in a way that isn't implemented now.

In short, there's likely an issue with your code that's setting off a recursive chain. Such bugs are quite tricky to track down.

I would start by eliminating pages from your build and using a slow introduction of pages to track down the culprit.

If it happens with no pages, check component autoloads or plugins.

**Comment by y1ann1s85** [comment]
Indeed, that is the approach we have taken. Trying to isolate the page/component that causes the error.

The tricky part is that we only get those errors in the production environment. In staging or local environment never. Everything works as expected. So the debugging process becomes even harder.

**Comment by pascalbaljet** [comment]
I'm closing this issue, as it's highly likely a memory leak in your app's codebase and not in Inertia itself. While you've already looked into `usePage`, memory leaks can come from various sources. If it seems random and not necessarily on a specific page, it could be some global loop or variable that's leaking.
